pkg/datasource.go

```diff
@@ -5,123 +5,478 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"math"
+	"regexp"
+	"time"
 
-	simplejson "github.com/bitly/go-simplejson"
+	"github.com/alexanderzobnin/grafana-zabbix/pkg/zabbix"
 	"github.com/grafana/grafana_plugin_model/go/datasource"
 	hclog "github.com/hashicorp/go-hclog"
-	plugin "github.com/hashicorp/go-plugin"
 )
 
-// ZabbixPlugin implements the Grafana backend interface and forwards queries to the ZabbixDatasource
-type ZabbixPlugin struct {
-	plugin.NetRPCUnsupportedPlugin
-	logger          hclog.Logger
-	datasourceCache *Cache
+// ZabbixDatasource stores state about a specific datasource and provides methods to make
+// requests to the Zabbix API
+type ZabbixDatasource struct {
+	client ZabbixAPIInterface
+	dsInfo *datasource.DatasourceInfo
+	hash   string
+	logger hclog.Logger
 }
 
-func (p *ZabbixPlugin) NewZabbixDatasource(dsInfo *datasource.DatasourceInfo) (*ZabbixDatasource, error) {
-	ds, err := newZabbixDatasource(dsInfo)
+// NewZabbixDatasource returns an instance of ZabbixDatasource with an API Client
+func NewZabbixDatasource(logger hclog.Logger, dsInfo *datasource.DatasourceInfo) (*ZabbixDatasource, error) {
+	client, err := NewZabbixAPIClient(logger, dsInfo.GetUrl())
```

Comment:
Something goes wrong here - ZabbixAPIClient required datasource set for correct work, but ZabbixDatasource requred ZabbixAPIClient. Looks like circular dependency.

Reply:
Yeah, it's definitely a circular reference. It was a compromise to keep the changes you made that remove dsInfo from all of the function parameters. A great improvement but not something that we resolved in our refactor.

Need to double check what ZabbixAPIClient needs from ZabbixDatasource. If it's just dsInfo, maybe we can store that on the client instead.